Adding additional core/user store selectors

ryanwelcher opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Addresses issue #1176

## Relevant technical choices

This PR is branched from the PR for #1175 which will need to be merged to `develop` first.

<!-- Please describe your changes. -->

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

felixarntz commented on assets/js/googlesitekit/datastore/user/user-info.js:
```diff
@@ -0,0 +1,212 @@
+	 */
+	getUser( state ) {
+		const { user, verified } = state;
+		return { user, verified };
```
This is a bit odd I think - this selector should rather just return the `user` object.

felixarntz commented on assets/js/googlesitekit/datastore/user/user-info.js:
```diff
@@ -0,0 +1,212 @@
+	 * @return {(boolean|undefined)} The user ID.
+	 */
+	isVerified: createRegistrySelector( ( select ) => () => {
+		const { verified } = select( STORE_NAME ).getUser() || {};
```
This value should come directly from the datastore state instead (no need for `createRegistrySelector` here). What this will need though (based on the other comment that `getUser` should not include `verified`) is a resolver that is essentially the same as the resolver for `getUser`, just checking for the `isVerified` selector instead.

felixarntz commented on assets/js/googlesitekit/datastore/user/user-info.js:
```diff
@@ -0,0 +1,212 @@
+	*getUser() {
+		const { select } = yield Data.commonActions.getRegistry();
+
+		const { user, verified } = select( STORE_NAME ).getUser();
```
Based on the other comments, this only should receive `user`, and shouldn't need to worry about `verified`.

Basically, both `getUser` and `isVerified` resolvers would result in the same logic, but since the check for whether a value is already set, that shouldn't be a problem. Plus it's not making an API request, so no real overhead anyway.

felixarntz commented on includes/Core/Authentication/Authentication.php:
```diff
@@ -586,6 +586,21 @@ private function inline_js_admin_data( $data ) {
+				$user['user'] = $new_data;
+				return $user;
+			}
+		);
```
Let's decouple this filter hook from `inline_js_admin_data` (since that is legacy) and hook it in in the `register` method. The logic here regarding getting the (Google account) profile data can be duplicated, that's fine in this case (legacy vs non-legacy). We also don't need to use `wp_get_current_user`, let's only update the parts of the `$user['user']` data that need to be updated (`email` and `picture`, if set).

felixarntz commented on includes/Core/Authentication/Authentication.php:
```diff
@@ -633,6 +648,15 @@ private function inline_js_setup_data( $data ) {
+				$user['verified'] = $data['isVerified'];
+				return $user;
+			}
+		);
```
Same here, let's move this to the `register` method and duplicate the relevant logic, since `inline_js_setup_data` is legacy code that this shouldn't be tied into.

ryanwelcher commented:
@googlebot I fixed it.

ryanwelcher commented:
@felixarntz I addressed your feedback but just wanted to call out that I actually added an action ( and tests ) to handle when `isVerified()` is called. It's very much the same as `getUser` but I thought adding it in would provide the structure for future changes if needed.

On the PHP side, I moved the hook as instructed but just wanted to confirm that the user `email` and `picture` are coming from the WP user and not `$this->profile` in the `Authentication` class.

felixarntz commented on assets/js/googlesitekit/datastore/user/user-info.js:
```diff
@@ -0,0 +1,256 @@
+			return;
+		}
+
+		yield actions.receiveUserIsVerified( { ...global._googlesitekitUserData } );
```
I think this is not correct, it should only pass the `verified` bit as single parameter (no object).

felixarntz commented on assets/js/googlesitekit/datastore/user/user-info.js:
```diff
@@ -0,0 +1,256 @@
+			return;
+		}
+
+		yield actions.receiveUserInfo( { ...global._googlesitekitUserData } );
```
Regarding my previous comment, this should maybe become `actions.receiveUser( global._googlesitekitUserData.user )`.

felixarntz commented on includes/Core/Authentication/Authentication.php:
```diff
@@ -260,6 +260,21 @@ function () {
+				$user['verified'] = $this->verification->has();
+				return $user;
+			}
+		);
```
👍

felixarntz commented on assets/js/googlesitekit/datastore/user/user-info.js:
```diff
@@ -0,0 +1,256 @@
+	 */
+	receiveUserInfo( userInfo ) {
+		invariant( userInfo, 'userInfo is required.' );
+		const { user, verified } = userInfo;
```
Or alternatively we all bring it back into one (that is how the `core/site` store has it) - the key change we'd need to make from how that was before here though would be that we would add a `getUserInfo` selector`, and that would have the single necessary resolver that populates all data from the global. Then the `getUser` selector would select from `getUserInfo`, as would `isVerified`.

ryanwelcher commented:
@felixarntz I went with the approach of keeping them separate. Feedback addressed.
